internal/pkg/encryption/encryption.go

```diff
@@ -230,46 +261,129 @@ func (h *Handler) syncKeys(k *encryption.Key, path string) error {
-	if valid, err := h.encryptionProvider.CheckKey(path, newKey); err != nil {
+func (h *Handler) updateKey(path string, existingKey *encryption.Key, handler keys.Handler) error {
+	valid, err := h.checkKey(path, handler)
+	if err != nil {
 		return err
-	} else if !valid {
-		// re-add the key to the slot
-		err = h.encryptionProvider.RemoveKey(path, newKey.Slot, existingKey)
-		if err != nil {
-			return fmt.Errorf("failed to drop old key during key update %w", err)
-		}
+	}
 
-		err = h.encryptionProvider.AddKey(path, existingKey, newKey)
-		if err != nil {
-			return fmt.Errorf("failed to add new key during key update %w", err)
+	if valid {
+		return nil
+	}
+
+	// re-add the key to the slot
+	err = h.encryptionProvider.RemoveKey(path, handler.Slot(), existingKey)
+	if err != nil {
+		return fmt.Errorf("failed to drop old key during key update %w", err)
+	}
+
+	err = h.addKey(path, existingKey, handler)
+	if err != nil {
+		return fmt.Errorf("failed to add new key during key update %w", err)
+	}
+
+	return err
+}
+
+func (h *Handler) checkKey(path string, handler keys.Handler) (bool, error) {
+	token, err := h.readToken(path, handler.Slot())
+	if err != nil {
+		return false, err
+	}
+
+	key, err := handler.GetKey(token)
+	if err != nil {
+		if errors.Is(err, keys.ErrTokenInvalid) {
+			return false, nil
 		}
 
+		return false, err
+	}
+
+	return h.encryptionProvider.CheckKey(path, key)
+}
+
+func (h *Handler) addKey(path string, existingKey *encryption.Key, handler keys.Handler) error {
+	key, token, err := handler.NewKey()
+	if err != nil {
 		return err
 	}
 
+	if token != nil {
+		if err = h.encryptionProvider.SetToken(path, key.Slot, token); err != nil {
+			return err
+		}
+	}
+
+	err = h.encryptionProvider.AddKey(path, existingKey, key)
+	if err != nil {
+		return fmt.Errorf("failed to add new key during key update %w", err)
+	}
+
 	return nil
 }
 
-func getKeys(encryptionConfig config.Encryption, partition *gpt.Partition) ([]*encryption.Key, error) {
-	encryptionKeys := make([]*encryption.Key, len(encryptionConfig.Keys()))
+func (h *Handler) initKeyHandlers(encryptionConfig config.Encryption, partition *gpt.Partition) ([]keys.Handler, error) {
+	handlers := make([]keys.Handler, 0, len(encryptionConfig.Keys()))
 
-	for i, cfg := range encryptionConfig.Keys() {
-		handler, err := keys.NewHandler(cfg)
+	for _, cfg := range encryptionConfig.Keys() {
+		handler, err := keys.NewHandler(cfg, keys.WithPartitionLabel(partition.Name))
 		if err != nil {
 			return nil, err
 		}
 
-		k, err := handler.GetKey(keys.WithPartitionLabel(partition.Name))
+		handlers = append(handlers, handler)
+	}
+
+	//nolint:scopelint
+	sort.Slice(handlers, func(i, j int) bool { return handlers[i].Slot() < handlers[j].Slot() })
+
+	return handlers, nil
+}
+
+func (h *Handler) tryHandlers(handlers []keys.Handler, cb func(h keys.Handler) error) error {
+	var errs error
+
+	for _, h := range handlers {
```

Comment:
should we have a retry loop here: we try each slot, if the unlock fails for all them, we do another loop?

if we fail to unlock disk encryption, there's no much we can do to proceed with the boot, so I think we should try hard even e.g. if the KMS is not available, or TPM gives back an error.

Reply:
yeah, good idea. Let me fix that